File: src/medium.cpp
#include "medium.h"
#include <new>
#include <cmath>
#include <cstdio>
#include <omp.h>

namespace batoid {

    #if defined(BATOID_GPU)
        #pragma omp declare target
    #endif


    ////////////
    // Medium //
    ////////////

    Medium::Medium() {}

    Medium::~Medium() {}


    /////////////////
    // ConstMedium //
    /////////////////

    ConstMedium::ConstMedium(const double n) : _n(n) {}

    ConstMedium::~ConstMedium() {}

    double ConstMedium::getN(const double wavelength) const {
        return _n;
    }


    /////////////////
    // TableMedium //
    /////////////////

    TableMedium::TableMedium(
        const double* args, const double* vals, const size_t size
    ) :
        Medium(), _args(args), _vals(vals), _size(size)
    {}

    TableMedium::~TableMedium() {}

    double TableMedium::getN(const double wavelength) const {
        // Linear search.  Better for GPU's?  and not that painful for small arrays?
        if (wavelength < _args[0])
            return NAN;
        if (wavelength > _args[_size-1])
            return NAN;
        int upperIdx;
        for(upperIdx=1; upperIdx<_size; upperIdx++) {
            if (wavelength < _args[upperIdx])
                break;
        }
        double out = (wavelength - _args[upperIdx-1]);
        out *= (_vals[upperIdx] - _vals[upperIdx-1]);
        out /= (_args[upperIdx] - _args[upperIdx-1]);
        out += _vals[upperIdx-1];
        return out;
    }


    /////////////////////
    // SellmeierMedium //
    /////////////////////

    SellmeierMedium::SellmeierMedium(
        double B1, double B2, double B3,
        double C1, double C2, double C3
    ) :
        Medium(), _B1(B1), _B2(B2), _B3(B3), _C1(C1), _C2(C2), _C3(C3)
    {}

    SellmeierMedium::~SellmeierMedium() {}

    double SellmeierMedium::getN(const double wavelength) const {
        // Sellmeier coefficients assume wavelength is in microns, so we have to multiply (1e6)**2
        double x = wavelength*wavelength*1e12;
        return std::sqrt(1.0 + _B1*x/(x-_C1) + _B2*x/(x-_C2) + _B3*x/(x-_C3));
    }


    /////////////////////
    // SumitaMedium //
    /////////////////////

    SumitaMedium::SumitaMedium(
        double A0, double A1, double A2,
        double A3, double A4, double A5
    ) :
        Medium(), _A0(A0), _A1(A1), _A2(A2), _A3(A3), _A4(A4), _A5(A5)
    {}

    SumitaMedium::~SumitaMedium() {}

    double SumitaMedium::getN(const double wavelength) const {
        //Sumita coefficients assume wavelength is in microns, so we have to multiply (1e6)**2
        double x = wavelength*wavelength*1e12;
        double y = 1./x;
        return std::sqrt(_A0 + _A1*x + y*(_A2 + y*(_A3 + y*(_A4 + y*_A5))));
    }


    /////////
    // Air //
    /////////

    // Uses the formulae given in Filippenko (1982), which appear to come from Edlen (1953),
    // and Coleman, Bozman, and Meggers (1960).  The units of the original formula are non-SI,
    // being mmHg for pressure (and water vapor pressure), and degrees C for temperature.  This
    // class accepts SI units, however, and transforms them when plugging into the formula.
    //
    // The default values for temperature, pressure and water vapor pressure are expected to be
    // appropriate for LSST at Cerro Pachon, Chile, but they are broadly reasonable for most
    // observatories.
    //
    // @param pressure       Air pressure in kiloPascals.
    // @param temperature    Temperature in Kelvins.
    // @param H2O_pressure   Water vapor pressure in kiloPascals.

    Air::Air(double pressure, double temperature, double h2o_pressure) :
        _pressure(pressure), _temperature(temperature), _h2o_pressure(h2o_pressure),
        _P(pressure * 7.50061683), _T(temperature - 273.15), _W(h2o_pressure * 7.50061683)
    {}

    Air::~Air() {}

    double Air::getN(const double wavelength) const {
        // inverse wavenumber squared in micron^-2
        double sigma_squared = 1e-12 / (wavelength*wavelength);
        double n_minus_one = (64.328 + (29498.1 / (146.0 - sigma_squared))
                                + (255.4 / (41.0 - sigma_squared))) * 1.e-6;
        n_minus_one *= _P * (1.0 + (1.049 - 0.0157 * _T) * 1.e-6 * _P) / (720.883 * (1.0 + 0.003661 * _T));
        n_minus_one -= (0.0624 - 0.000680 * sigma_squared)/(1.0 + 0.003661 * _T) * _W * 1.e-6;
        return 1+n_minus_one;
    }


    #if defined(BATOID_GPU)
        #pragma omp end declare target
    #endif


    //////////////////
    // MediumHandle //
    //////////////////

    MediumHandle::MediumHandle() :
        _hostPtr(nullptr),
        _devicePtr(nullptr)
    {}

    MediumHandle::~MediumHandle() {}

    const Medium* MediumHandle::getPtr() const {
        #if defined(BATOID_GPU)
            return _devicePtr;
        #else
            return _hostPtr;
        #endif
    }

    const Medium* MediumHandle::getHostPtr() const {
        return _hostPtr;
    }


    ///////////////////////
    // ConstMediumHandle //
    ///////////////////////

    ConstMediumHandle::ConstMediumHandle(const double n) :
        MediumHandle()
    {
        _hostPtr = new ConstMedium(n);
        #if defined(BATOID_GPU)
            auto alloc = omp_target_alloc(sizeof(ConstMedium), omp_get_default_device());
            #pragma omp target map(from:_devicePtr), is_device_ptr(alloc)
            {
                _devicePtr = new (alloc) ConstMedium(n);
            }
        #endif
    }

    ConstMediumHandle::~ConstMediumHandle() {
        #if defined(BATOID_GPU)
            omp_target_free(_devicePtr, omp_get_default_device());
        #endif
        delete _hostPtr;
    }


    ///////////////////////
    // TableMediumHandle //
    ///////////////////////

    TableMediumHandle::TableMediumHandle(const double* args, const double* vals, const size_t size) :
        MediumHandle(),
        _args(args), _vals(vals), _size(size)
    {
        _hostPtr = new TableMedium(args, vals, size);
        #if defined(BATOID_GPU)
            auto alloc = omp_target_alloc(sizeof(TableMedium), omp_get_default_device());
            const size_t lcl_size = _size;
            const double* lcl_args = _args;
            const double* lcl_vals = _vals;
            #pragma omp target enter data map(to:lcl_args[:lcl_size], lcl_vals[:lcl_size])
            #pragma omp target map(from:_devicePtr), is_device_ptr(alloc)
            {
                _devicePtr = new (alloc) TableMedium(lcl_args, lcl_vals, lcl_size);
            }
        #endif
    }

    TableMediumHandle::~TableMediumHandle() {
        #if defined(BATOID_GPU)
            #pragma omp target exit data \
                map(release:_args[:_size], _vals[:_size])
            omp_target_free(_devicePtr, omp_get_default_device());
        #endif
        delete _hostPtr;
    }


    ///////////////////////////
    // SellmeierMediumHandle //
    ///////////////////////////

    SellmeierMediumHandle::SellmeierMediumHandle(
        double B1, double B2, double B3,
        double C1, double C2, double C3
    ) :
        MediumHandle()
    {
        _hostPtr = new SellmeierMedium(B1, B2, B3, C1, C2, C3);
        #if defined(BATOID_GPU)
            auto alloc = omp_target_alloc(sizeof(SellmeierMedium), omp_get_default_device());
            #pragma omp target map(from:_devicePtr), is_device_ptr(alloc)
            {
                _devicePtr = new (alloc) SellmeierMedium(B1, B2, B3, C1, C2, C3);
            }
        #endif
    }

    SellmeierMediumHandle::~SellmeierMediumHandle() {
        #if defined(BATOID_GPU)
            omp_target_free(_devicePtr, omp_get_default_device());
        #endif
        delete _hostPtr;
    }


    ////////////////////////
    // SumitaMediumHandle //
    ////////////////////////

    SumitaMediumHandle::SumitaMediumHandle(
        double A0, double A1, double A2,
        double A3, double A4, double A5
    ) :
        MediumHandle()
    {
        _hostPtr = new SumitaMedium(A0, A1, A2, A3, A4, A5);
        #if defined(BATOID_GPU)
            auto alloc = omp_target_alloc(sizeof(SumitaMedium), omp_get_default_device());
            #pragma omp target map(from:_devicePtr), is_device_ptr(alloc)
            {
                _devicePtr = new (alloc) SumitaMedium(A0, A1, A2, A3, A4, A5);
            }
        #endif
    }

    SumitaMediumHandle::~SumitaMediumHandle() {
        #if defined(BATOID_GPU)
            omp_target_free(_devicePtr, omp_get_default_device());
        #endif
        delete _hostPtr;
    }

    AirHandle::AirHandle(double pressure, double temperature, double h2o_pressure) :
        MediumHandle()
    {
        _hostPtr = new Air(pressure, temperature, h2o_pressure);
        #if defined(BATOID_GPU)
            auto alloc = omp_target_alloc(sizeof(Air), omp_get_default_device());
            #pragma omp target map(from:_devicePtr), is_device_ptr(alloc)
            {
                _devicePtr = new (alloc) Air(pressure, temperature, h2o_pressure);
            }
        #endif
    }

    AirHandle::~AirHandle() {
        #if defined(BATOID_GPU)
            omp_target_free(_devicePtr, omp_get_default_device());
        #endif
        delete _hostPtr;
    }
}

File: batoid/medium.py
import numpy as np
from . import _batoid


class Medium:
    """The `Medium` class is used to model a refractive medium.

    Media have essentially one function: to return their refractive index given
    a wavelength.
    """
    def getN(self, wavelength):
        """Return refractive index.

        Parameters
        ----------
        wavelength : float
            Vacuum wavelength in meters.

        Returns
        -------
        n : float
            Refractive index.
        """
        return self._medium.getN(wavelength)

    def __ne__(self, rhs):
        return not (self == rhs)


class ConstMedium(Medium):
    """A `Medium` with wavelength-independent refractive index.

    Parameters
    ----------
    n : float
        The refractive index.
    """
    def __init__(self, n):
        self.n = n
        self._medium = _batoid.CPPConstMedium(n)

    def __eq__(self, rhs):
        if type(rhs) == type(self):
            return self.n == rhs.n
        return False

    def __getstate__(self):
        return self.n

    def __setstate__(self, n):
        self.n = n
        self._medium = _batoid.CPPConstMedium(n)

    def __hash__(self):
        return hash(("batoid.ConstMedium", self.n))

    def __repr__(self):
        return f"ConstMedium({self.n})"


class TableMedium(Medium):
    """A `Medium` with refractive index defined via a lookup table.

    Parameters
    ----------
    wavelengths : array of float
        Wavelengths in meters.
    ns : array of float
        Refractive indices.
    """
    def __init__(self, wavelengths, ns):
        self.wavelengths = np.array(wavelengths)
        self.ns = np.array(ns)
        self._medium = _batoid.CPPTableMedium(
            self.wavelengths.ctypes.data,
            self.ns.ctypes.data,
            len(self.wavelengths)
        )

    @classmethod
    def fromTxt(cls, filename, **kwargs):
        """Load a text file with refractive index information in it.
        The file should have two columns, the first with wavelength in microns,
        and the second with the corresponding refractive indices.
        """
        import os
        try:
            wavelength, n = np.loadtxt(filename, unpack=True, **kwargs)
        except IOError:
            import glob
            from . import datadir
            filenames = glob.glob(os.path.join(datadir, "**", "*.txt"))
            for candidate in filenames:
                if os.path.basename(candidate) == filename:
                    wavelength, n = np.loadtxt(candidate, unpack=True, **kwargs)
                    break
            else:
                raise FileNotFoundError(filename)
        return TableMedium(wavelength*1e-6, n)

    def __eq__(self, rhs):
        if type(rhs) == type(self):
            return (
                np.array_equal(self.wavelengths, rhs.wavelengths)
                and np.array_equal(self.ns, rhs.ns)
            )
        return False

    def __getstate__(self):
        return self.wavelengths, self.ns

    def __setstate__(self, args):
        self.__init__(*args)

    def __hash__(self):
        return hash((
            "batoid.TableMedium", tuple(self.wavelengths), tuple(self.ns)
        ))

    def __repr__(self):
        return f"TableMedium({self.wavelengths!r}, {self.ns!r})"


class SellmeierMedium(Medium):
    r"""A `Medium` with Sellmeier dispersion formula.

    The Sellmeier formula is

    .. math::

        n = \sqrt{1 + \sum_{i=1}^3 \frac{B_i \lambda^2}{\lambda^2 - C_i}}

    where :math:`\lambda` is the vacuum wavelength in microns.

    Parameters
    ----------
    coefs: array of float
        Sellmeier coefficients (B1, B2, B3, C1, C2, C3)
    """
    def __init__(self, *args, **kwargs):
        if len(args) == 6:
            coefs = tuple(args)
        elif len(args) == 1:
            coefs = tuple(args[0])
        elif kwargs:
            coefs = tuple([
                kwargs[k] for k in ['B1', 'B2', 'B3', 'C1', 'C2', 'C3']
            ])
        else:
            raise ValueError("Incorrect number of arguments")
        self.coefs = coefs
        self._medium = _batoid.CPPSellmeierMedium(*coefs)

    def __eq__(self, rhs):
        if type(rhs) == type(self):
            return self.coefs == rhs.coefs
        return False

    def __getstate__(self):
        return self.coefs

    def __setstate__(self, coefs):
        self.coefs = coefs
        self._medium = _batoid.CPPSellmeierMedium(*coefs)

    def __hash__(self):
        return hash(("batoid.SellmeierMedium", self.coefs))

    def __repr__(self):
        return f"SellmeierMedium({self.coefs})"


class SumitaMedium(Medium):
    r"""A `Medium` with Sumita dispersion formula, also known as the Schott
    dispersion formula.

    The Sumita formula is

    .. math::

        n = \sqrt{A_0 + A_1 \lambda^2 + \sum_{i=2}^5 A_i \lambda^{-2 (i-1)}}

    where :math:`\lambda` is the vacuum wavelength in microns.

    Parameters
    ----------
    coefs: array of float
        Sumita coefficients (A0, A1, A2, A3, A4, A5)
    """
    def __init__(self, *args, **kwargs):
        if len(args) == 6:
            coefs = tuple(args)
        elif len(args) == 1:
            coefs = tuple(args[0])
        elif kwargs:
            coefs = tuple([
                kwargs[k] for k in ['A0', 'A1', 'A2', 'A3', 'A4', 'A5']
            ])
        else:
            raise ValueError("Incorrect number of arguments")
        self.coefs = coefs
        self._medium = _batoid.CPPSumitaMedium(*coefs)

    def __eq__(self, rhs):
        if type(rhs) == type(self):
            return self.coefs == rhs.coefs
        return False

    def __getstate__(self):
        return self.coefs

    def __setstate__(self, coefs):
        self.coefs = coefs
        self._medium = _batoid.CPPSumitaMedium(*coefs)

    def __hash__(self):
        return hash(("batoid.SumitaMedium", self.coefs))

    def __repr__(self):
        return f"SumitaMedium({self.coefs})"


class Air(Medium):
    """A `Medium` for air.

    Parameters
    ----------
    pressure : float, optional
        Atmospheric pressure in kiloPascals.  [default: 69.328]
    temperature : float, optional
        Temperature in Kelvin.  [default: 293.15]
    h2o_pressure : float, optional
        Water vapor pressure in kiloPascals.  [default: 1.067]

    Notes
    -----
    Uses the formulae given in Filippenko (1982), which appear to come from
    Edlen (1953), and Coleman, Bozman, and Meggers (1960).  The default values
    for temperature, pressure and water vapor pressure are expected to be
    appropriate for LSST at Cerro Pachon, Chile, but they are broadly
    reasonable for most observatories.
    """
    def __init__(self, pressure=69.328, temperature=293.15, h2o_pressure=1.067):
        self.pressure = pressure
        self.temperature = temperature
        self.h2o_pressure = h2o_pressure
        self._medium = _batoid.CPPAir(pressure, temperature, h2o_pressure)

    def __eq__(self, rhs):
        if type(rhs) == type(self):
            return (
                self.pressure == rhs.pressure
                and self.temperature == rhs.temperature
                and self.h2o_pressure == rhs.h2o_pressure
            )
        return False

    def __getstate__(self):
        return self.pressure, self.temperature, self.h2o_pressure

    def __setstate__(self, args):
        self.pressure, self.temperature, self.h2o_pressure = args
        self._medium = _batoid.CPPAir(
            self.pressure, self.temperature, self.h2o_pressure
        )

    def __hash__(self):
        return hash((
            "batoid.Air", self.pressure, self.temperature, self.h2o_pressure
        ))

    def __repr__(self):
        return f"Air({self.pressure}, {self.temperature}, {self.h2o_pressure})"
